rfc/001-truncated-block-history.md

```diff
@@ -0,0 +1,87 @@
+# RFC 001: Coordination and Configuration of Block Retention and State Sync
+
+## Changelog
+
+- 2020-03-23: Initial draft (@erikgrinaker)
+
+## Author(s)
+
+- Erik Grinaker (@erikgrinaker)
+
+## Context
+
+Currently, all Tendermint nodes contain the complete sequence of blocks from genesis up to some height (typically the latest chain height). This will no longer be true when the following features are released:
+
+* [Block pruning](https://github.com/tendermint/tendermint/issues/3652): removes historical blocks and associated data (e.g. validator sets) up to some height, keeping only the most recent blocks.
+
+* [State sync](https://github.com/tendermint/tendermint/issues/828): bootstraps a new node by syncing the state machine at a given height, but not historical blocks and associated data.
+
+To maintain the integrity of the chain, the use of these features must be coordinated such that necessary historical blocks will not be made unavailable or lost forever. In particular:
+
+* Some nodes should have complete block histories, for auditability and querying.
+
+* The majority of nodes should retain blocks longer than the Cosmos SDK unbonding period, for light client verification.
+
+* Some nodes must take and serve state sync snapshots with snapshot intervals less than the block retention periods, to allow new nodes to state sync and then replay blocks to catch up.
+
+* Only a minority of nodes can be state synced within the unbonding period, for light client verification and to serve block histories for catch-up.
+
+However, it is unclear if and how we should enforce this. It may not be possible to technically enforce all of these without knowing the state of the entire network, but it may also be unrealistic to expect this to be enforced entirely through social consensus. This is especially unfortunate since the consequences of misconfiguration can be permanent chain-wide data loss.
+
+The main configuration options involved are:
+
+* Unbonding time (SDK): the duration in which validators can be economically punished for misbehavior.
+
+* Block time (Tendermint): the minimum duration between consecutive blocks.
+
+* Block retention (Tendermint): the number of recent blocks (heights) to retain.
+
+* Snapshot interval (SDK): the interval (in heights) between taking state sync snapshots.
```

Comment:
The nomenclature will get confusing for operators WRT to data backend pruning strategies. I'd just add a small note here to make a clear distinguishment between what snapshots are and what pruning snapshots are.

Reply:
I share that concern. Any ideas for alternatives?

Reply:
maybe pruning checkpoints?

Reply:
Yeah, I think that makes sense. I’ll open a separate issue/PR.

Reply:
https://github.com/tendermint/iavl/issues/229